waku/v2/node/storage/message/waku_message_store.nim

```diff
@@ -211,6 +266,156 @@ method getAll*(db: WakuMessageStore, onData: message_store.DataProc): MessageSto
+  trace "dbPageSize adjusted to: ",  ret
+  ret
+
+
+method getPage*(db: WakuMessageStore,
+              pred: QueryFilterMatcher,
+              pagingInfo: PagingInfo):
+             MessageStoreResult[(seq[WakuMessage], PagingInfo, HistoryResponseError)] =
+  ## Get a single page of history matching the predicate and
+  ## adhering to the pagingInfo parameters
+  
+  trace "getting page from SQLite DB", pagingInfo=pagingInfo
+
+  let
+    responsePageSize = if pagingInfo.pageSize == 0 or pagingInfo.pageSize > MaxPageSize: MaxPageSize # Used default MaxPageSize for invalid pagingInfos
+                  else: pagingInfo.pageSize
+
+  var dbPageSize = responsePageSize  # we retrieve larger pages from the DB for queries with (sparse) filters (TODO: improve adaptive dbPageSize increase)
+
+  var cursor = pagingInfo.cursor
+
+  var messages: seq[WakuMessage]
+  var
+    lastIndex: Index
+    numRecordsVisitedPage: uint64 = 0  # number of DB records visited during retrieving the last page from the DB
+    numRecordsVisitedTotal: uint64 = 0 # number of DB records visited in total
+    numRecordsMatchingPred: uint64 = 0 # number of records that matched the predicate on the last DB page; we use this as to gauge the sparseness of rows matching the filter.
+
+  proc msg(s: ptr sqlite3_stmt) = # this is the actual onData proc that is passed to the query proc (the message store adds one indirection)
+    if uint64(messages.len) >= responsePageSize: return
+    let
+      receiverTimestamp = column_timestamp(s, 0)
+
+      topic = cast[ptr UncheckedArray[byte]](sqlite3_column_blob(s, 1))
+      topicLength = sqlite3_column_bytes(s,1)
+      contentTopic = ContentTopic(string.fromBytes(@(toOpenArray(topic, 0, topicLength-1))))
+
+      p = cast[ptr UncheckedArray[byte]](sqlite3_column_blob(s, 2))
+      length = sqlite3_column_bytes(s, 2)
+      payload = @(toOpenArray(p, 0, length-1))
+
+      pubsubTopicPointer = cast[ptr UncheckedArray[byte]](sqlite3_column_blob(s, 3))
+      pubsubTopicLength = sqlite3_column_bytes(s,3)
+      pubsubTopic = string.fromBytes(@(toOpenArray(pubsubTopicPointer, 0, pubsubTopicLength-1)))
+
+      version = sqlite3_column_int64(s, 4)
+
+      senderTimestamp = column_timestamp(s, 5)
+      retMsg = WakuMessage(contentTopic: contentTopic, payload: payload, version: uint32(version), timestamp: Timestamp(senderTimestamp))
+      # TODO: we should consolidate WakuMessage, Index, and IndexedWakuMessage; reason: avoid unnecessary copying and recalculation
+      index = retMsg.computeIndex(receiverTimestamp, pubsubTopic) # TODO: retrieve digest from DB
+      indexedWakuMsg = IndexedWakuMessage(msg: retMsg, index: index, pubsubTopic: pubsubTopic) # TODO: constructing indexedWakuMsg requires unnecessary copying
+
+    lastIndex = index
+    numRecordsVisitedPage += 1
+    try:
+      if pred(indexedWakuMsg): #TODO throws unknown exception
+        numRecordsMatchingPred += 1
+        messages.add(retMsg)
```

Comment:
This is on my todo list :).  It would make the code cleaner as we would not need the adaptive DB pages anymore.
This would, however, limit the expressiveness of the predicate.
If we want to offer an easy to use interface to the store, we have to translate the nim closure into a WHERE compatible predicate. 
For now, I think there are larger performance gains in other areas (e.g. caching), so I postponed this.